libexec/rtld-elf/aarch64/rtld_machdep.h

```diff
@@ -36,23 +37,145 @@
 #include <sys/types.h>
 #include <machine/atomic.h>
 
+#if __has_feature(capabilities)
+#include <cheri/cheri.h>
+#include <cheri/cheric.h>
+#endif
+
 struct Struct_Obj_Entry;
 
 /* Return the address of the .dynamic section in the dynamic linker. */
+#ifdef __CHERI_PURE_CAPABILITY__
+#define rtld_dynamic(obj) (&_DYNAMIC)
+#else
 #define	rtld_dynamic(obj)						\
 ({									\
 	Elf_Addr _dynamic_addr;						\
 	asm volatile("adr	%0, _DYNAMIC" : "=&r"(_dynamic_addr));	\
 	(const Elf_Dyn *)_dynamic_addr;					\
 })
+#endif
 
 Elf_Addr reloc_jmpslot(Elf_Addr *where, Elf_Addr target,
     const struct Struct_Obj_Entry *defobj, const struct Struct_Obj_Entry *obj,
     const Elf_Rel *rel);
 
+#if __has_feature(capabilities)
+
+#define	FUNC_PTR_REMOVE_PERMS						\
+	(CHERI_PERM_SEAL | CHERI_PERM_STORE | CHERI_PERM_STORE_CAP |	\
+	CHERI_PERM_STORE_LOCAL_CAP)
+
+#define	DATA_PTR_REMOVE_PERMS						\
+	(CHERI_PERM_SEAL | CHERI_PERM_EXECUTE)
+
+/* TODO: we should have a separate member for .text/rodata */
+#define get_codesegment(obj) ((obj)->text_rodata_cap)
+
+/* TODO: ABIs with tight bounds */
+#define can_use_tight_pcc_bounds(obj) ((void)(obj), false)
+
+#define set_bounds_if_nonnull(ptr, size)	\
+	do { if (ptr) { ptr = cheri_setbounds_sametype(ptr, size); } } while(0)
+
+#endif /* __has_feature(capabilities) */
+
+#if __has_feature(capabilities)
+
+typedef void (* __capability dlfunccap_t)(struct __dlfunc_arg);
+
+/*
+ * Create a pointer to a function.
+ */
+static inline dlfunccap_t
+make_code_cap_pointer(const Elf_Sym *def, const struct Struct_Obj_Entry *defobj,
+    bool tight_bounds, size_t addend)
+{
+#ifdef __CHERI_PURE_CAPABILITY__
+	const void *ret = get_codesegment(defobj) + def->st_value;
+#else
+	const void * __capability ret = cheri_setoffset(cheri_getdefault(),
+	    (uintptr_t)(defobj->relocbase + def->st_value));
+#endif
```

Comment:
Perhaps get_codesegment could instead be defined as `cheri_getdefault() + (vaddr_t)obj->relocbase` on hybrid? There's no need for the cheri_setoffset though as DDC should have 0 offset (well, in some Morello configs it won't matter, but the kernel will always ensure it's 0 on exec, and if userspace wants to mess itself up then it can go for it)